src/Tasks/LargeFileUploadTask.php

```diff
@@ -0,0 +1,345 @@
+
+        $uploadUrl = $this->getValidatedUploadUrl($this->uploadSession);
+
+        if (empty($uploadUrl)) {
+            throw new InvalidArgumentException('The upload session URL must not be empty.');
+        }
+        $requestInformation->setUri($uploadUrl);
+        return  $this->adapter->sendNoContentAsync($requestInformation)
+                              ->then(function ($result) {
+                                      if (method_exists($this->uploadSession, 'setIsCancelled')){
+                                          $this->uploadSession->setIsCancelled(true);
+                                      }
+                                      else if (method_exists($this->uploadSession, 'setAdditionalData') && method_exists($this->uploadSession, 'getAdditionalData')){
+                                          $current = $this->uploadSession->getAdditionalData();
+                                          $new = array_merge($current, ['isCancelled' => true]);
+                                          $this->uploadSession->setAdditionalData($new);
+                                      }
+                                      return $this->uploadSession;
+                              });
+    }
+
+    /**
+     * @param Parsable $parsable
+     * @param array<string> $propertyCandidates
+     * @return array{boolean,mixed}
+     */
+    private function additionalDataContains(Parsable $parsable, array $propertyCandidates): array  {
+        if (!is_subclass_of($parsable, AdditionalDataHolder::class)) {
+            throw new InvalidArgumentException('The object passed does not contain propert(y|ies) ['.implode(',',$propertyCandidates).'] and does not implement AdditionalDataHolder');
+        }
+        $additionalData = $parsable->getAdditionalData();
+        foreach ($propertyCandidates as $propertyCandidate) {
+            if (isset($additionalData[$propertyCandidate])) {
+                return [true, $additionalData[$propertyCandidate]];
+            }
+        }
+        return [false, null];
+    }
+
+    /**
+     * @param Parsable $parsable
+     * @param string $getterName
+     * @param array<string> $propertyNamesInAdditionalData
+     * @return array{bool, mixed}
+     */
+    private function checkValueExists(Parsable $parsable, string $getterName, array $propertyNamesInAdditionalData): array {
+        $checkedAdditionalData = $this->additionalDataContains($parsable, $propertyNamesInAdditionalData);
+        if (is_subclass_of($parsable, AdditionalDataHolder::class) && $checkedAdditionalData[0]) {
+            return [true, $checkedAdditionalData[1]];
+        }
+
+        if (method_exists($parsable, $getterName)) {
+            return [true, $parsable->{$getterName}()];
+        }
+        return [false, null];
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function resume(Parsable $uploadSession, ?callable $onRangeUploadComplete = null): void {
```

Comment:
Shouldn't it resume with the upload session & callback that had already been initialised?